### PR TITLE
feat(ui5-popup): аdd default responsive paddings

### PR DESCRIPTION
**Pull request by LidiyaGeorgieva**
Default responsive paddings are added for all popups.
In the controls that they are not needed they are removed.
Sample with way to remove the paddings by the user added in the test pages.

Affected components: Dialog, Popover, ResponsivePopover, ColorPalettePopover, WizardPopover, BreadcrumbsPopover, DatePickerPopover, ViewSettingsDialog, ComboBox Suggestion popover, Input, File Uploader, MultiComboBox, Select, TabContainer, TimePicker, Tokenizer, BarcodeScanner, Breadcrumbs, ShellBar

Fixes: #4402

**Thank you for your contribution!** 👏

To get it merged faster, kindly review the checklist below:

## Pull Request Checklist
- [ ] Reviewed the [Contributing Guidelines](https://github.com/SAP/ui5-webcomponents/blob/master/CONTRIBUTING.md)
    + Especially the [How to Contribute](https://github.com/SAP/ui5-webcomponents/blob/master/CONTRIBUTING.md#how-to-contribute) section 
- [ ] [Correct commit message style](https://github.com/SAP/ui5-webcomponents/blob/master/docs/Guidelines.md#commit-message-style)

### CODE REVIEW

**Review by MapTo0 — CHANGES_REQUESTED** [review]
Do we really need to expose those parts? I am okay with that just asking if we need to make them public. Did you check IE?

**Comment by LidiyaGeorgieva** [comment]
> Do we really need to expose those parts? I am okay with that just asking if we need to make them public. Did you check IE?

We don't need to expose those parts. It's just one way to do it (removing the default responsive paddings). 
The other way is to provide class, that can be used by the application.
We should decide which approach is better.

Removing of the paddings in IE does not work if "parts" are used for it.

File: packages/main/src/themes/ResponsivePopoverCommon.css [hunk 1]
```diff
@@ -205,3 +205,7 @@
 	position: absolute;
 	right: 1rem;
 }
+
+.ui5-tab-container-responsive-popover::part(content) {
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/ResponsivePopoverCommon.css:209]: Move this to TabInOverflow.css

File: packages/main/src/themes/Dialog.css [hunk 2]
```diff
@@ -67,4 +69,30 @@
 	left: -0.25rem;
 	right: unset;
 	cursor: sw-resize;
-}
\ No newline at end of file
+}
+
+.ui5-popup-footer-root {
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/Dialog.css:74]: Move it to PopupsCommon.css

File: packages/main/src/themes/PopupsCommon.css [hunk 3]
```diff
@@ -85,3 +84,23 @@
...
+:host([media-range="M"]) .ui5-popup-content,
+:host([media-range="L"]) .ui5-popup-content {
+	padding: var(--_ui5_popup_content_padding_m_l);
+}
+
+:host([media-range="XL"]) .ui5-popup-content {
+	padding: var(--_ui5_popup_content_padding_xl);
+}
+
+/* ** Class to remove responsive paddings of the content **
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/PopupsCommon.css:102]: Remove this.

File: packages/main/src/themes/PopupsCommon.css [hunk 4]
```diff
@@ -85,3 +84,23 @@
 :host([disable-scrolling]) .ui5-popup-content {
     overflow: hidden;
 }
+
+/*** Responsive paddings for the content***/
+:host([media-range="S"]) .ui5-popup-content {
+	padding: var(--_ui5_popup_content_padding_s);
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/PopupsCommon.css:90]: padding: 1rem var(--_ui5_popup_content_padding_s);

File: packages/main/src/themes/PopupsCommon.css [hunk 5]
```diff
@@ -85,3 +84,23 @@
...
+}
+
+:host([media-range="M"]) .ui5-popup-content,
+:host([media-range="L"]) .ui5-popup-content {
+	padding: var(--_ui5_popup_content_padding_m_l);
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/PopupsCommon.css:95]: padding: 1rem var(--_ui5_popup_content_padding_m_l);

File: packages/main/src/themes/PopupsCommon.css [hunk 6]
```diff
@@ -85,3 +84,23 @@
...
+}
+
+:host([media-range="XL"]) .ui5-popup-content {
+	padding: var(--_ui5_popup_content_padding_xl);
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/PopupsCommon.css:99]: padding: 1rem var(--_ui5_popup_content_padding_xl);

File: packages/main/src/themes/base/Dialog-parameters.css [hunk 7]
```diff
@@ -6,4 +6,7 @@
 	--_ui5_dialog_header_border_radius: 0px;
 	--_ui5_dialog_resize_handle_right: -0.25rem;
 	--_ui5_dialog_resize_handle_bottom: -0.0625rem;
+	--_ui5_popup_header_footer_padding_s: 1rem;
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/base/Dialog-parameters.css:9]: Remove these

File: packages/main/src/themes/base/PopupsCommon-parameters.css [hunk 8]
```diff
@@ -1,5 +1,7 @@
 :root {
-	--_ui5_popup_content_padding: .4375em;
+	--_ui5_popup_content_padding_s: 1rem;
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/base/PopupsCommon-parameters.css:2]: --_ui5_popup_content_padding_s: 1rem;
--_ui5_popup_content_padding_m_l: 2rem;
--_ui5_popup_content_padding_xl: 3rem;

File: packages/main/src/themes/sap_horizon/Dialog-parameters.css [hunk 9]
```diff
@@ -8,4 +8,7 @@
 	--_ui5_dialog_resize_handle_right: 1px;
 	--_ui5_dialog_header_focus_offset: -2px;
 	--_ui5_dialog_header_focus_width: 2px;
+	--_ui5_popup_header_footer_padding_s: 1rem;
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/sap_horizon/Dialog-parameters.css:11]: Remove these

File: packages/main/src/themes/Dialog.css [hunk 10]
```diff
@@ -67,4 +69,30 @@
...
+	min-height: 2.5rem;
+	justify-content: end;
+}
+
+/*** Responsive paddings for the Header and Footer ***/
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/Dialog.css:79]: Move all these to PopupsCommon.css

File: packages/main/src/themes/sap_horizon/PopupsCommon-parameters.css [hunk 11]
```diff
@@ -1,7 +1,9 @@
 @import "../base/PopupsCommon-parameters.css";
 
 :root {
-	--_ui5_popup_content_padding: .4375em;
+	--_ui5_popup_content_padding_s: 1rem;
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/sap_horizon/PopupsCommon-parameters.css:4]: Remove these.

**Review by TeodorTaushanov — CHANGES_REQUESTED** [review]

File: packages/main/src/themes/Dialog.css [hunk 12]
```diff
@@ -22,6 +22,8 @@
 
 .ui5-popup-header-root {
```
Comment (TeodorTaushanov) [inline packages/main/src/themes/Dialog.css:23]: Move it to PopupCommons.css.

**Review by TeodorTaushanov — CHANGES_REQUESTED** [review]

**Review by TeodorTaushanov — APPROVED** [review]

**Review by vladitasev — CHANGES_REQUESTED** [review]
Please don't use tags in CSS as they are not scoping friendly. 
use the
```css
[ui5-dialog]
```
form instead.

**Comment by nnaydenow** [comment]
ui5-select changes look good to me!

**Review by vladitasev — COMMENTED** [review]
Looks good, will checkout tomorrow morning and can +1 afterwards

**Review by vladitasev — CHANGES_REQUESTED** [review]
Currently, you create 2 ResizeHandlers for each Popover/Dialog. One is created internally in Popup.js, and then another one for Popover/Dialog. Both are executed when you call `super.onEnterDOM`. 

Please rework as follows:
 - only attach ResizeHandlers once in Popup.js
 - the `_resize` function in `Popup.js` will be the only resize handler callback
 - In `Dialog.js` / `Popover.js` extend the `_resize` function to call `super._resize` first, and then add the dialog/popover specific logic there. Also, for clarity, annotate the `_resize` function in the 2 derivativa classes with `@override`.

File: packages/main/test/pages/Popups.html [hunk 13]
```diff
@@ -25,7 +25,7 @@
 
 	<link rel="stylesheet" type="text/css" href="styles/Popups.css">
 </head>
-<body class="popups1auto">
+<body class="popups1auto" onload="onload()">
```
Comment (MapTo0) [inline packages/main/test/pages/Popups.html:28]: I would add `defer` to the script, move it on the bottom of the page or attach to the load event in the `popups.js` file rather than inlining this. Up to you, just a suggestion.

**Review by MapTo0 — DISMISSED** [review]
Inputs look broken on mobile. Check:
- https://sap.github.io/ui5-webcomponents/master/playground/main/pages/Input/
- http://localhost:8080/test-resources/pages/Input.html

Same is valid for MultiComboBox, ComboBox, MultiInput (tokenizer's, "n more" popover).

I see that most of the controls are using the Popups without paddings and this feature is mainly for apps. Did you. consider providing an internal API for disabling the feature? Having that would also lower the resize handlers count and might impact performance. In general, implementation looks good just check the inputs. Scenario with 100 inputs on a page would attach 100 Resize handlers and none of them is going to be used as the input needs 0 padding all the time.

**Comment by LidiyaGeorgieva** [comment]
> 
> I see that most of the controls are using the Popups without paddings and this feature is mainly for apps. Did you. consider providing an internal API for disabling the feature? Having that would also lower the resize handlers count and might impact performance. In general, implementation looks good just check the inputs. Scenario with 100 inputs on a page would attach 100 Resize handlers and none of them is going to be used as the input needs 0 padding all the time.

Dialog and Popover uses the Resize Handler for their positioning not only for setting the paddings of the content.

The Suggestion popover on mobile and Popups sample will be fixed.

**Review by TeodorTaushanov — DISMISSED** [review]

**Comment by ilhan007** [comment]
Some merge conflicts should be resolved

**Comment by nnaydenow** [comment]
Hi colleagues,

Did you investigate the @MapTo0's proposal? If we consider to implement it will allow us to remove these parts that were recently added and in addition all these padding changes will be maintained in a single place.

Regards,
Nayden

**Comment by LidiyaGeorgieva** [comment]
> Some merge conflicts should be resolved

Done

**Review by dimovpetar — APPROVED** [review]

File: packages/main/src/SelectPopover.hbs [hunk 14]
```diff
@@ -65,8 +65,7 @@
 	<ui5-popover
 		skip-registry-update
 		_disable-initial-focus
-		prevent-focus-restore
-		no-padding
+		prevent-focus-restoreg
```
Comment (MapTo0) [inline packages/main/src/SelectPopover.hbs:68]: restoreg?

**Comment by MapTo0** [comment]
MultiInput's tokenizer's dialog (opened when you click N more on mobile) still has paddings.

**Review by tsanislavgatev — CHANGES_REQUESTED** [review]
ViewSettingsDialog header padding is now too big and it pushes the bar content too much into the middle of the control.
The rest of the controls on our side look fine.

**Comment by MapTo0** [comment]
> Yes, we investigate and discuss the @MapTo0's proposal.
> Before the start of the implementation there was a long discussion about the approach (on mail, with people from different teams).
> "Parts" was the chosen one since it is more flexible - there is more use-cases than just remove all default paddings (remove left, right, set fixed paddings, etc.). This flexibility will be useful not only in the web components, but also in the applications.

Having specific left / right padding can be controlled in the specific components, but IMHO having or not paddings in general can be handled in the popups (and every components sets e.g. left padding if needed). However, I am fine with the change so far. Looks good to me.

Small diffs I manage to find testing against master:

- ComboBox and Input (on mobile) after the change the footer of the Dialog has some paddings. Is this desired behaviour?
- ShellBar's popover's paddings (when items are truncated on smaller screens are wrong) @nnaydenow check please
- Wizard's Progress Navigator Steps popover on small screens - paddings are broken after the change

Basically all controls with predefined paddings are still broken and you have to check all components in the library.
Please do more testing and go trough all components.

**Comment by LidiyaGeorgieva** [comment]
ComboBox and Input (on mobile) after the change the footer of the Dialog has some paddings. Is this desired behaviour?
ShellBar's popover's paddings (when items are truncated on smaller screens are wrong) @nnaydenow check please
Wizard's Progress Navigator Steps popover on small screens - paddings are broken after the change

> ViewSettingsDialog header padding is now too big and it pushes the bar content too much into the middle of the control. The rest of the controls on our side look fine.

**Comment by LidiyaGeorgieva** [comment]
> Small diffs I manage to find testing against master:
> 
> * ComboBox and Input (on mobile) after the change the footer of the Dialog has some paddings. Is this desired behaviour?
> * ShellBar's popover's paddings (when items are truncated on smaller screens are wrong) @nnaydenow check please
> * Wizard's Progress Navigator Steps popover on small screens - paddings are broken after the change

 ComboBox and Input (on mobile) -> Paddings should be responsive for the header, footer and the content. The target is all elements to be vertically aligned.

 ShellBar's popover's paddings -> Why wrong? Before there was 7px padding (--_ui5_popup_content_padding, which was not by design), now it is the new one (responsive)
 
 Wizard's Progress Navigator Steps popover -> same as ShellBar

**Comment by nnaydenow** [comment]
Hi @LidiyaGeorgieva,

Sorry for inconvenience. Could you please recheck BarcodeScanner, Breadcrumbs, ShellBar components.

Regards,
Nayden

**Comment by LidiyaGeorgieva** [comment]
> BarcodeScanner

Thank you for noticing that. I checked them. For all three of them is the same:
Before there was 7px padding (--_ui5_popup_content_padding, which was not by design), now they have the new one (responsive)

**Review by dimovpetar — DISMISSED** [review]
- ColorPalettePopover has extra padding now in the header, when on mobile. Check if its own can be removed, and only use the one from the dialog
- BarCodeScanner has extra padding in the footer, check if its own can be removed

**Comment by LidiyaGeorgieva** [comment]
> * ColorPalettePopover has extra padding now in the header, when on mobile. Check if its own can be removed, and only use the one from the dialog
> * BarCodeScanner has extra padding in the footer, check if its own can be removed

Done

File: packages/main/src/themes/PopupsCommon.css [hunk 15]
```diff
@@ -85,3 +81,47 @@
 :host([disable-scrolling]) .ui5-popup-content {
     overflow: hidden;
 }
+
+/*** Responsive paddings for the content***/
+:host([media-range="S"]) .ui5-popup-content {
+	padding: 1rem var(--_ui5_popup_content_padding_s);
```
Comment (gmkv) [inline packages/main/src/themes/PopupsCommon.css:87]: These responsive paddings for the content are defined as:
Size S: 0 1rem;
Size M: 0 2rem;
Size L: 0 2rem;
Size XL: 0 3rem;

Remove the top and bottom paddings.

File: packages/main/src/themes/ValueStateMessage.css [hunk 16]
```diff
@@ -74,3 +74,16 @@
...
+	padding: 0; 
+}
+
+.ui5-valuestatemessage-popover::part(header) {
+	min-height: 0;
+}
+
+.ui5-valuestatemessage-popover::part(footer) {
+	min-height: 0;
+}
```
Comment (gmkv) [inline packages/main/src/themes/ValueStateMessage.css:89]: can you combine these rules?

File: packages/main/test/specs/Dialog.spec.js [hunk 17]
```diff
@@ -378,3 +378,54 @@ describe("Page scrolling", () => {
...
+		const actualHeaderPadding = await dialog.shadow$(".ui5-popup-header-root").getCSSProperty("padding-left");
+
+		// footer
+		const actualFooterPadding = await dialog.shadow$(".ui5-popup-footer-root").getCSSProperty("padding-left");
+
+		assert.strictEqual(actualContentPadding.value, expectedPadding, "dialog has correct padding set on the content");
+		assert.strictEqual(actualHeaderPadding.value, expectedPadding, "dialog has correct padding set on the header");
+		assert.strictEqual(actualFooterPadding.value, expectedPadding, "dialog has correct padding set on the footer");
+
+		browser.$("#btnCloseDialog").click();
```
Comment (gmkv) [inline packages/main/test/specs/Dialog.spec.js:407]: await this click as well

**Review by gmkv — CHANGES_REQUESTED** [review]

File: packages/main/src/themes/PopupsCommon.css (hunk 15 above)
Comment (LidiyaGeorgieva) [reply packages/main/src/themes/PopupsCommon.css:87]: Why to remove top and bottom paddings?
See: B2 paddings here https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2033847432

Comment (gmkv) [reply packages/main/src/themes/PopupsCommon.css:87]: Ok, this is fine then. Keep the top and bottom padding.

**Review by gmkv — APPROVED** [review]
